src/RealizationOrder.cpp

```diff
@@ -13,14 +16,98 @@ using std::set;
 using std::vector;
 using std::pair;
 
+namespace {
+
+class CheckLoopCarriedDependence : public IRVisitor {
+private:
+    const string &func;
+    const vector<Expr> &prev_args;
+
+    using IRVisitor::visit;
+
+    void visit(const Call *op) {
+        if ((op->call_type == Call::Halide) && (func == op->name)) {
+            internal_assert(prev_args.size() == op->args.size())
+                << "Self-reference should have the same number of args as the original\n";
+            for (size_t i = 0; i < op->args.size(); i++) {
+                if (!equal(op->args[i], prev_args[i])) {
+                    debug(4) << "Self-reference of " << op->name << " with different args "
+                             << "from the LHS. Encounter loop-carried dependence.\n";
+                    result = false;
+                    return;
+                }
+            }
+        }
+    }
+
+public:
+    bool result;
+
+    CheckLoopCarriedDependence(const string &f, const vector<Expr> &prev_args)
+        : func(f), prev_args(prev_args), result(true) {}
+};
+
+// Check to see if there is a loop-carried dependence in a 'func' update
+// definition with the previous definition it's computed with.
+bool has_loop_carried_dependence(const string &func, const vector<Expr> &prev_args,
```

Comment:
I'm concerned about this case:

f(x) = x;
RDom r(0, 3);
f(r) += 3;
f(r+1) *= 2;

Fusing on r changes the output